tests/test_frame.py

```diff
@@ -1251,3 +1252,13 @@ def test_boolean_loc() -> None:
     df = pd.DataFrame([[0, 1], [1, 0]], columns=[True, False], index=[True, False])
     check(assert_type(df.loc[True], pd.Series), pd.Series)
     check(assert_type(df.loc[:, False], pd.Series), pd.Series)
+
+
+def test_groupby_result() -> None:
+    # GH 142
+    df = pd.DataFrame({"a": [0, 1, 2], "b": [4, 5, 6], "c": [7, 8, 9]})
+    lresult = [(cast(Tuple[int, int], k), g) for k, g in df.groupby(["a", "b"])]
```

Comment:
I would be very cautious adding `cast`s to the type tests, if we add them we might in end up testing (only) the runtime API of pandas
```py
index, value = assert_type(df.groupby(["a", "b"]), tuple[Hashable, DataFrame]
assert isinstance(index.__hash__(), int)  # if possible isinstance(index, Hashable) 
assert isinstance(value, DataFrame)
```

Reply:
I did something like this in the next commit.  `cast` has been removed.  It's a bit tricky to test `__iter__()`

Reply:
Changed.

Also did a scheme to address the original issue.